_posts/jared_baker/2018-12-15-Bluestore.md

```diff
@@ -0,0 +1,171 @@
+
+### Draining
+To drain a node we would modify the osd crush weight of all of a storage nodes OSD's to 0. This will trigger an evacuation of data away from this storage node and distribute it to the rest of the cluster.
+
+Example, to drain OSD's 720 to 755 we would run:
+~~~bash
+for i in $(seq 720 755); do ceph osd crush reweight osd.$i 0; done
+~~~
+
+When the draining is complete you can check ceph's health (ceph -s) to confirm if it is done draining the data. You can also use 'ceph osd df tree' to see how much data the disks still have to transfer off. Our storage nodes took on average 24 hours per server to drain.
+
+### Converting the OSD to Bluestore
+Follow these steps to safely retire the OSD and bring it back as Bluestore.
+[Available in script form here](https://github.com/CancerCollaboratory/infrastructure/blob/master/migrate_bluestore.sh)
+
+~~~bash
+# Stop the OSD process
+systemctl stop ceph-osd@<osd-id>.service
+
+# Unmount the OSD
+umount /dev/<block-device>
+
+# Zap the disk
+ceph-disk zap <osd-id>
+
+# Mark the OSD as destroyed
+ceph osd destroy <osd-id> --yes-i-really-mean-it
+
+# Prepare the disk as Bluestore
+ceph-disk prepare --bluestore /dev/<block-device> --osd-id <osd-id>
+~~~
+
+Notice that when comparing the output of ceph osd tree before/after conversion that the class column may now indicate your osd type (ssd or hdd) which for us, was blank before.
+
+Filestore 'ceph osd tree'
+~~~bash
+ID   CLASS WEIGHT     TYPE NAME                STATUS REWEIGHT PRI-AFF
+  -1       8503.16504 root default
+ -84                0     rack 1
+  -5       1306.07776     rack rack1
+ -83        388.79956         host storage1-r1                               
+720         10.79999             osd.720         up  1.00000 1.00000
+721         10.79999             osd.721         up  1.00000 1.00000
+~~~
+
+Bluestore 'ceph osd tree'
+~~~bash
+ID   CLASS WEIGHT     TYPE NAME                STATUS REWEIGHT PRI-AFF
+  -1       8503.16504 root default
+ -84                0     rack 1
+  -5       1306.07776     rack rack1
+ -83        388.79956         host storage1-r1
+720   hdd   10.79999             osd.720         up  1.00000 1.00000
+721   hdd   10.79999             osd.721         up  1.00000 1.00000
+~~~
+
+### Filling
+Now it's time to populate the bluestore OSD's with data. Just like the draining step we will change the osd crush weight
+
+Example, to fill OSD's 720 to 755 (12TB disks) we would run:
+~~~bash
+for i in $(seq 720 755); do ceph osd crush reweight osd.$i 10.79999; done
+~~~
+
+Depending on the size of the disks you will need to find the right value for your osd crush weight. This is typically the size of the disk in TB. Read more about the differences of ceph osd reweight and ceph osd crush reweight [HERE](https://ceph.com/geen-categorie/difference-between-ceph-osd-reweight-and-ceph-osd-crush-reweight/)
```

Comment:
- Dot at the end of the sentence
- I would not capitalize HERE, but that's personal preferences